devices/nand-imx6ull/nand.c

```diff
@@ -0,0 +1,682 @@
+			blen = 0;
+		}
+		else {
+			blen -= 8 - boffs;
+		}
+		ret += 8 - __builtin_popcount(data8);
+	}
+
+	/* Check bytes until 32-bit aligned address */
+	while ((blen > 8) && ((addr_t)buff8 % sizeof(data32))) {
+		data8 = *buff8++;
+		blen -= 8;
+		ret += 8 - __builtin_popcount(data8);
+	}
+
+	/* Check 32-bit words */
+	buff32 = (const u32 *)buff8;
+	while (blen > 8 * sizeof(data32)) {
+		data32 = *buff32++;
+		blen -= 8 * sizeof(data32);
+
+		if (data32 == 0xffffffff) {
+			continue;
+		}
+		ret += 8 * sizeof(data32) - __builtin_popcount(data32);
+	}
+
+	/* Check rest of the bytes */
+	buff8 = (const u8 *)buff32;
+	while (blen > 8) {
+		data8 = *buff8++;
+		blen -= 8;
+		ret += 8 - __builtin_popcount(data8);
+	}
+
+	/* Check last byte */
+	if (blen > 0) {
+		data8 = *buff8;
+		data8 |= (u8)(0xff >> blen);
+		ret += 8 - __builtin_popcount(data8);
+	}
+
+	return ret;
+}
+
+
+static int nand_checkECC(const nand_t *nand, unsigned int page, u8 *data, u8 *aux, unsigned int chunks)
+{
+	const nand_dev_t *dev = (const nand_dev_t *)nand;
+	const unsigned int mlen = 8 * dev->cfg->metasz + dev->ecc->ecc0 * dev->ecc->gf0;
+	const unsigned int dlen = 8 * dev->ecc->blockszN + dev->ecc->eccN * dev->ecc->gfN;
+	u8 *status = aux + (dev->cfg->metasz + sizeof(u32) - 1) / sizeof(u32), *raw = NULL;
+	unsigned int i, boffs, blen, flips;
+	int err;
+
+	for (i = 0; i < chunks; i++) {
+		/* BCH reported uncorrectable chunk */
+		if (status[i] == 0xfe) {
+			/* Check for an erased chunk with bitflips */
+			/* Read raw page data and count bitflips manually */
+			if (raw == NULL) {
+				raw = nand_buff;
+				err = nand_read(nand, page, raw, NULL, 1);
+				if (err < 0) {
+					return err;
+				}
+			}
+
+			/* Metadata chunk */
+			if (i == 0) {
+				boffs = 0;
+				blen = mlen;
+			}
+			/* Data chunk */
+			else {
+				boffs = mlen + (i - 1) * dlen;
+				blen = dlen;
+			}
+
+			flips = nand_checkErased(raw, boffs, blen);
+			if (flips == 0) {
+				continue;
+			}
+
+			/* Handle metadata bitflips */
+			if (i == 0) {
+				/* Too many metadata bitflips */
+				if (flips > dev->ecc->ecc0) {
+					return -EFAULT;
+				}
+
+				/* Correct metadata chunk */
+				hal_memset(aux, 0xff, dev->cfg->metasz);
+			}
+			/* Handle data chunk bitflips */
+			else {
+				if (flips > dev->ecc->eccN) {
+					return -EFAULT;
+				}
+
+				/* Correct data chunk */
+				hal_memset(data + (i - 1) * dev->ecc->blockszN, 0xff, dev->ecc->blockszN);
+			}
+		}
+	}
+
+	return EOK;
+}
+
+
+int nand_read(const nand_t *nand, unsigned int page, void *data, void *aux, int raw)
+{
+	const nand_dev_t *dev = (const nand_dev_t *)nand;
+	dma_t *dma = (dma_t *)nand_dma;
+	u8 addr[NAND_MAX_ADDRSZ] = { 0 };
+	u16 size;
+	int err;
+
+	/* Prepare page address (skip 2 bytes for column address) */
+	hal_memcpy(addr + 2, &page, sizeof(addr) - 2);
+
+	/* Read whole page */
+	if (data != NULL) {
+		size = dev->cfg->pagesz + dev->cfg->oobsz;
+		/* Read without ECC */
+		if (raw) {
```

Comment:
```suggestion
		if (raw != NULL) {
```

Reply:
Applies to other cases of pointer checking

Reply:
`raw` is a boolean value, not a pointer.

Reply:
Sorry, my mistake. if (raw != 0) then ;) (to avoid implicit int -> bool cast, raw is still an int)

Reply:
Fixed.